impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -47,140 +59,134 @@ public class KafkaClusterTest {
 
     private static final System.Logger LOGGER = System.getLogger(KafkaClusterTest.class.getName());
     private static final Duration CLUSTER_FORMATION_TIMEOUT = Duration.ofSeconds(10);
-    private TestInfo testInfo;
-    private KeytoolCertificateGenerator keytoolCertificateGenerator;
-
-    @Test
-    public void kafkaClusterKraftMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(true)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static int clusterSizesIndex = 0;
+
+    private static Stream<Version> versions() {
+        return Stream.of(
+                version("latest")
+        // TODO: waiting for new versions support in ozangunalp repo https://github.com/ozangunalp/kafka-native/issues/21
+        // version("3.3.1"),
+        // version("3.2.1")
+        );
     }
 
-    @Test
-    public void kafkaClusterZookeeperMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static Stream<BrokerCluster> clusterSizes() {
+        return Stream.of(
+                brokerCluster(1),
+                brokerCluster(2));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterKraftMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(true)
-                .build())) {
-            assumeTrue(cluster instanceof TestcontainersKafkaCluster, "KAFKA-14287: kraft timing out on shutdown in multinode case");
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
-        }
+    private final Object[] brokerClusters = clusterSizes().toArray();
+
+    static Stream<List<Annotation>> clusterSizesAndModes() {
+        return Stream.of(
+                List.of(brokerCluster(1), kraftCluster(1)),
+                List.of(brokerCluster(2), kraftCluster(1)),
+                List.of(brokerCluster(1), zooKeeperCluster()),
+                List.of(brokerCluster(2), zooKeeperCluster()));
```

Comment:
I think I understand what this is doing, but it feels like a clunky way to express the problem. 

As I understand it we want to express a range of Kafka cluster sizes and a set of controller types and then generate the cartesian product of those two things. 

Which isn't quite what the parameter resolver does, at least I don't think that's what it does. I feel like we are working at the wrong level of abstraction because we manually duplicate the `kraftCluster` and the `zooKeeperCluster`. 

@tombentley what was your vision here? Am I missing something? 

Should we be using two `@DimensionMethodSource`s instead?

Reply:
I agree with you, but here I did not found a way to use `@DimensionMethodSource` that to switch between `kraftCluster` and `zookeeperCluster`, because returning an `Stream<Annotation>` does not work for `@DimensionMethodSource`.

So I decided to use the `@ConstraintsMethodSource` to avoid rewriting the test twice, one per each cluster mode.

My vision using the templates is a tool to reduce writing tests with `@Test` that test exactly the same but the configuration is different and can be done using the annotations.

Reply:
Moved all template tests to `TemplateTest` class